### PR TITLE
File transfer upload and download merge, file list

### PR DESCRIPTION
**Pull request by biochill**
- Upload + Download dialogs into one "New Transfer" dialog
- File list view

### CODE REVIEW

File: yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto [hunk 1]
```diff
@@ -208,16 +233,13 @@ message CreateTransferRequest {
...
   
-  optional DownloadOptions downloadOptions = 6;
+  optional UploadOptions uploadOptions = 6;
   
-  // Configuration options specific to ``UPLOAD`` transfers.
-  optional UploadOptions uploadOptions = 7;
-
   //used to derive the source entity id
-  optional string source = 9;
+  optional string source = 7;
```
Comment (fqqb) [inline yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto:239]: Could we keep this at 9?
(Renumbering will break backwards compatibility in Python client)

File: yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto [hunk 2]
```diff
@@ -208,16 +233,13 @@ message CreateTransferRequest {
...
 
   //used to derive the destination entity id
-  optional string destination = 10;
+  optional string destination = 9;
```
Comment (fqqb) [inline yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto:242]: Could we keep this at 10?
(Renumbering will break backwards compatibility in Python client)

File: yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto [hunk 3]
```diff
@@ -76,6 +77,29 @@ service FileTransferApi {
...
+    option (yamcs.api.route) = {
+      post: "/api/filetransfer/{instance}/{serviceName}/requestfiles"
+      body: "*"
+    };
+  }
+
+  // Get latest file list
+  rpc GetFileList(ListFilesRequest) returns (ListFilesResponse) {
+    option (yamcs.api.route) = {
+      post: "/api/filetransfer/{instance}/{serviceName}/getfiles"
```
Comment (fqqb) [inline yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto:99]: This seems like it should be a get method. Suggest this:

`get: "/api/filetransfer/{instance}/{serviceName}/files"`

Then also get rid of `body: "*"`

File: yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto [hunk 4]
```diff
@@ -76,6 +77,29 @@ service FileTransferApi {
...
+  rpc SubscribeRemoteFileList(SubscribeTransfersRequest) returns (stream ListFilesResponse) {
+    option (yamcs.api.websocket) = {
+      topic: "remote-file-list"
+    };
+  }
+
+  // Request file list from remote
+  rpc RequestFileList(ListFilesRequest) returns (google.protobuf.Empty) {
+    option (yamcs.api.route) = {
+      post: "/api/filetransfer/{instance}/{serviceName}/requestfiles"
```
Comment (fqqb) [inline yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto:91]: Verbs in the path should be separated by colon, I suggest maybe this:
`post: "/api/filetransfer/{instance}/{serviceName}/files:sync"`

File: yamcs-web/src/main/webapp/src/app/shared/forms/ObjectSelector.ts [hunk 5]
```diff
@@ -55,17 +55,32 @@ export class ObjectSelector implements ControlValueAccessor, OnChanges, OnDestro
...
   }
 
+  // Called when breadcrumb is selected
   changePrefix(prefix: string) {
     this.loadCurrentFolder(prefix);
   }
 
   private loadCurrentFolder(prefix?: string) {
+    if (!this.bucket) {
+      // The object selector for remote files does not have a bucket.
```
Comment (fqqb) [inline yamcs-web/src/main/webapp/src/app/shared/forms/ObjectSelector.ts:72]: Think I would prefer that ObjectSelector remains purely used for selecting bucket objects.
Maybe make a RemoteFileSelector as part of the FileTransferModule?

Comment (fqqb) [inline yamcs-web/src/main/webapp/src/app/shared/forms/ObjectSelector.ts:72]: Think I would prefer that ObjectSelector remains purely used for selecting bucket objects.
Maybe make a RemoteFileSelector as part of the FileTransferModule?

**Review by fqqb — COMMENTED** [review]
Thanks. See inline comments

File: yamcs-core/src/main/java/org/yamcs/http/api/FileTransferApi.java [hunk 6]
```diff
@@ -163,6 +166,13 @@ public void createTransfer(Context ctx, CreateTransferRequest request, Observer<
             transferOptions.setOverwrite(true);
             transferOptions.setCreatePath(true);
 
+			if (request.hasUploadOptions()) {
```
Comment (fqqb) [inline yamcs-core/src/main/java/org/yamcs/http/api/FileTransferApi.java:169]: sry few more tabs here

File: yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto [hunk 7]
```diff
@@ -208,11 +232,8 @@ message CreateTransferRequest {
   // **Required** The path at the remote entity. Example: ``a/local/path/some_filename``.
   optional string remotePath = 5;
   
-  optional DownloadOptions downloadOptions = 6;
+  optional UploadOptions uploadOptions = 6;
```
Comment (fqqb) [inline yamcs-api/src/main/proto/yamcs/protobuf/filetransfer/filetransfer.proto:235]: Can you also revert uploadOptions back to its original serialization number 7?
(same reason as before -- i hadn't noticed it before)

**Review by fqqb — COMMENTED** [review]
ty. Two minor comments. Else OK as far as I'm concerned.

**Comment by biochill** [comment]
I have fixed your issues and then some more. Let me know if you find anything else.

**Pull request closed without merging**